af/shovel/centrifugation.py

```diff
@@ -1273,6 +1273,46 @@ def pop(datum):
         for key in known_keys:
             t.pop(key)
 
+class TelegramFeeder(BaseFeeder):
+    min_compat_code_ver = 6
+    data_table = sink_table = 'telegram'
+    columns = ('msm_no', 'telegram_web_failure', 'telegram_web_blocking',
+               'telegram_http_blocking', 'telegram_tcp_blocking')
+
+    @staticmethod
+    def row(msm_no, datum):
+        if datum['test_name'] != 'telegram':
+            return ''
+        t = datum['test_keys']
+
+        web_failure = t.get('telegram_web_failure', None)
+        web_status = t.get('telegram_web_status', None)
+        web_blocking = None
+        if web_status == 'ok':
+            web_blocking = False
+        elif web_status == 'blocked':
+            web_blocking = True
+
+        http_blocking = t.get('telegram_http_blocking', None)
+        tcp_blocking = t.get('telegram_tcp_blocking', None)
+        return '{:d}\t{}\t{}\t{}\t{}\n'.format(
+                    msm_no,
+                    pg_quote(web_failure),
+                    pg_quote(web_blocking),
+                    pg_quote(http_blocking),
+                    pg_quote(tcp_blocking)
+        )
+
+    @staticmethod
+    def pop(datum):
+        if datum['test_name'] != 'telegram':
+            return
+        t = datum['test_keys']
+        known_keys = ('telegram_web_failure', 'telegram_http_blocking',
+                      'telegram_web_status', 'telegram_tcp_blocking')
+        for key in known_keys:
+            t.pop(key)
```

Comment:
`web_status` values besides `ok` and `blocked` are not parsed, but the key is pop'ed despite being potentially malformed. That may be a bad idea depending on circumstances.

Reply:
What is the best practice for handling cases of malformed data? Should I be throwing an exception?

Reply:
IIRC, I tried to follow the following pattern: if the key has unknown (malformed?) value then the code does not `pop` it. So it ends in `residual` and marks the measurement as a measurement deserving some attention within a processes discussed in #137.

Reply:
I see. And how do handle checking the validity of the value given that the conversion logic goes inside of the getter?

Reply:
Other feeders follow the pattern when the every present field is validated at `row()` and an exception (or assertion) is raised if the field has an invalid value.